### PR TITLE
Estimate price tests

### PR DESCRIPTION
**Pull request by maurolacy**
Some integration tests of the `EstimatePrice` query.

These tests require the "sender" account to have funds, so that price estimation can proceed... the "estimation" also sends coins, that is, (afaik) **it performs the swap** between the sender account and the pool module.

There are a number of issues here:

- Queries shouldn't change state / generate transactions.
- Sender is an arbitrary string field in the message. That means this "query" could be used (afaik) to **perform swaps on behalf of other people's accounts**.

### CODE REVIEW

**Comment by ethanfrey** [comment]
Note that the actual query code cannot write to the state, thanks to https://github.com/CosmWasm/wasmd/pull/611/files

Please note the `.CacheContext()` added in https://github.com/CosmWasm/wasmd/pull/611/files which is still in `master`. This means any writes performed on a query plugin will be dropped before returning the result to the calling contract.

Not ideal behavior from the gamm module, but x/wasm already deals with this

**Comment by maurolacy** [comment]
Good to know, thanks, makes sense.

What about this query point being called directly, i.e. thorough a grpc / REST call?

Not directly relevant to our case, but good to know about / check that behaviour nevertheless I think.

**Comment by ethanfrey** [comment]
> What about this query point being called directly, i.e. thorough a grpc / REST call?

On incoming requests, there are 3 different states router to - simulate, checktx, and delivertx (the actual one called for execution). query works off a read-only cache of the last block state.

(read only means this is a read-write cache over the last block that is never written to disk and dropped after each query)

**Comment by ethanfrey** [comment]
Let me know when this is ready to review

**Comment by maurolacy** [comment]
It is already. Will just remove the FIXMEs / Improve docs.

File: app/wasm/queries.go [hunk 1]
```diff
@@ -71,7 +71,10 @@ func (qp QueryPlugin) EstimatePrice(ctx sdk.Context, estimatePrice *bindings.Est
 		return nil, wasmvmtypes.UnsupportedRequest{Kind: "TODO: multi-hop swap price estimation"}
 	}
 
-	var senderAddress = sdk.AccAddress(sender)
+	senderAddress, err := sdk.AccAddressFromBech32(sender)
```
Comment (ethanfrey) [inline app/wasm/queries.go:74]: Yes, good catch. The old wouldn't work at all

File: app/wasm/test/custom_query_test.go [hunk 2]
```diff
@@ -173,6 +173,96 @@ func TestQuerySpotPrice(t *testing.T) {
...
+
+	// Estimate swap rate
+	uosmo, err := poolFunds[0].Amount.ToDec().Float64()
+	require.NoError(t, err)
+	ustar, err := poolFunds[1].Amount.ToDec().Float64()
+	require.NoError(t, err)
+	swapRate := ustar / uosmo
+
+	// Query estimate cost (Exact in. No route)
+	amountIn := sdk.NewInt(1)
```
Comment (ethanfrey) [inline app/wasm/test/custom_query_test.go]: Why 1?
No one ever swaps 0.000001 OSMO - less than gas fees.

Let's use normal numbers like 1.000.000 and this also limits the issues with rounding errors

File: app/wasm/test/custom_query_test.go [hunk 3]
```diff
@@ -173,6 +173,96 @@ func TestQuerySpotPrice(t *testing.T) {
 	require.InEpsilonf(t, expected+swapFee, price, epsilon, fmt.Sprintf("Outside of tolerance (%f)", epsilon))
 }
 
+func TestQueryEstimatePrice(t *testing.T) {
+	actor := RandomAccountAddress()
+	osmosis, ctx := SetupCustomApp(t, actor)
+	epsilon := 1. // high epsilon (because of rounding?)
```
Comment (ethanfrey) [inline app/wasm/test/custom_query_test.go]: That is too high. The answer could be between 0 and 40 and it would be considered correct.

Please use a normal swap amount, like 1 million (1 OSMO) and then set epsilon to 0.01. If it is more than 1% off we have a serious problem (and should likely be off much less)

File: app/wasm/test/custom_query_test.go [hunk 4]
```diff
@@ -173,6 +173,96 @@ func TestQuerySpotPrice(t *testing.T) {
...
+	amount, err := amountIn.ToDec().Float64()
+	require.NoError(t, err)
+	expected := amount * swapRate // out
+	require.InEpsilonf(t, expected, cost, epsilon, fmt.Sprintf("Outside of tolerance (%f)", epsilon))
+
+	// And the other way around
+	// Query estimate cost (Exact out. No route)
+	amountOut := sdk.NewInt(1)
+	query = wasmbindings.OsmosisQuery{
+		EstimatePrice: &wasmbindings.EstimatePrice{
```
Comment (ethanfrey) [inline app/wasm/test/custom_query_test.go:235]: Reflecting on this, we are not estimating the Price, as the output is the payout (not a decimal price like in spot price). Maybe EstimateSwap would be a better name?

(Making this rename is out of scope of the PR and needs a new osmosis-bindings release, but I want your opinion on it)

File: app/wasm/test/custom_query_test.go [hunk 5]
```diff
@@ -173,6 +173,96 @@ func TestQuerySpotPrice(t *testing.T) {
...
+			Contract: reflect.String(),
+			First: wasmbindings.Swap{
+				PoolId:   starPool,
+				DenomIn:  "ustar",
```
Comment (ethanfrey) [inline app/wasm/test/custom_query_test.go]: How about leaving it the same order (swapping uosmo for star, but setting the output.)

Set output to eg. 20 million and expect around 1 million

File: app/wasm/test/custom_query_test.go (hunk 3 above)
Comment (ethanfrey) [reply app/wasm/test/custom_query_test.go]: Ah, I forgot about slippage.
Let's either trade 10.000 (so it is much less than pool funds).
Or multiply the pool funds by 100 so we have small slippage.

File: app/wasm/test/custom_query_test.go [hunk 6]
```diff
@@ -173,6 +173,96 @@ func TestQuerySpotPrice(t *testing.T) {
...
+		},
+	}
+	resp = wasmbindings.EstimatePriceResponse{}
+	queryCustom(t, ctx, osmosis, reflect, query, &resp)
+	require.NotNil(t, resp.Amount.In)
+	require.Nil(t, resp.Amount.Out)
+	cost, err = (*resp.Amount.In).ToDec().Float64()
+	require.NoError(t, err)
+
+	uosmo, err = poolFunds[0].Amount.ToDec().Float64()
```
Comment (ethanfrey) [inline app/wasm/test/custom_query_test.go]: This can be removed, right? Cut and paste error?
We are using swapRate from above.

**Review by ethanfrey — COMMENTED** [review]
The flow looks good. epsilon is way too high.

Please increase the numbers being swapped to reasonable values and increase pool size as needed and then you should be able to reduce epsilon. At that point, these tests would make good sense.

I would also like a second test that runs the exact same query two times and ensure it is the exact same amount (an amount that was big enough to make some slippage, showing if this wrote to state in the query, then the numbers would be different, and the test shows we do not mutate state - this is in x/wasm, but good to demo it here... along to a link to the x/wasm PR)

File: app/wasm/test/custom_query_test.go (hunk 4 above)
Comment (maurolacy) [reply app/wasm/test/custom_query_test.go:235]: I agree. Let's refactor in another PR, after a new osmosis-bindings.

File: app/wasm/test/custom_query_test.go (hunk 5 above)
Comment (maurolacy) [reply app/wasm/test/custom_query_test.go]: OK.

**Comment by ethanfrey** [comment]
nice updates
